applications/notifications/dal.py

```diff
@@ -0,0 +1,346 @@
+    notify = None
+
+    if notification_id is not None:
+        query = query.filter(GroupNotification.id == notification_id)
+        notify = query.one_or_none()
+    else:
+        if group_id is not None:
+            query = query.filter(GroupNotification.group_id == group_id)
+
+        if resource is not None:
+            query = query.filter(GroupNotification.resource == resource)
+
+        if type is not None:
+            query = query.filter(GroupNotification.type == type)
+
+        if created_at is not None:
+            query = query.filter(GroupNotification.created_at <= datetime.strptime(
+                created_at, "%Y-%m-%dT%H:%M:%S.%f"))
+
+        if first_created_at is not None:
+            query = query.filter(GroupNotification.created_at >= datetime.strptime(
+                first_created_at, "%Y-%m-%dT%H:%M:%S.%f"))
+
+    query = query.filter(GroupNotification.is_read.is_(False))
+    total = query.count()
+    query = query.update({'is_read': True}, synchronize_session=False)
+
+    return total, notify
+
+
+# Get notification count per group for all groups
+@with_session
+def get_notification_count_per_group(db_session, member_email, is_read=None):
+
+    # Get notifications by merging similar ones
+    cte_query = db_session.query(GroupNotification.type,
+                                 GroupNotification.resource,
+                                 GroupNotification.is_read,
+                                 GroupNotification.member_email,
+                                 GroupNotification.created_at,
+                                 GroupNotification.group_id,
+                                 (func.row_number().over(order_by=desc(GroupNotification.created_at)) - func.row_number().over(partition_by=GroupNotification.type, order_by=desc(GroupNotification.created_at))).label('row_number'))
+
+    cte_query = cte_query.filter(
+        GroupNotification.member_email.ilike(member_email))
+
+    if is_read is not None:
+        cte_query = cte_query.filter(GroupNotification.is_read.is_(is_read))
+
+    cte_query = cte_query.cte('group_notification_sub_count_cte')
+
+    query = db_session.query(cte_query.c.type.label('type'),
+                             cte_query.c.resource.label('resource'),
+                             cte_query.c.group_id.label('group_id')) \
+        .group_by(cte_query.c.type, cte_query.c.row_number, cte_query.c.resource, cte_query.c.group_id) \
+        .order_by(desc(func.max(cte_query.c.created_at)))
+
+    total = query.count()
+    query = query.cte('group_notification_count_cte')
+
+    subquery = db_session.query(
+        query.c.group_id.label('group_id'), func.count(query.c.group_id).label('count'))
+
+    subquery = subquery.group_by(query.c.group_id).subquery()
+
+    new_query = db_session.query(Group, subquery.c.count).join(
+        subquery, subquery.c.group_id == Group.id).order_by(subquery.c.count.desc())
+
+    return total, new_query.all()
+
+    """
```

Comment:
Okay, since it is the same logic, but simplier/shorter, I think it is okay to leave it in for reference. Please just add a comment explaining that it is a simplified version of the query.